spec/03-plugins/14-request-transformer/02-access_spec.lua

```diff
@@ -653,6 +730,39 @@ for _, strategy in helpers.each_strategy() do
     end)
 
     describe("replace", function()
+
+      describe("Host header", function()
+        it("preserve_host = true, host change in plugin: transformed host", function()
+          local res = assert(proxy_client:send {
+            method  = "GET",
+            path    = "/get",
+            headers = {
+              ["Content-Type"] = "application/json",
+              host             = "preserve-host-replace-host.test"
+            }
+          })
+          assert.response(res).has.status(200)
+          local json = assert.response(res).has.jsonbody()
+          local value = assert.has.header("host", json)
+          assert.equals("replaced", value)
+        end)
+
+        it("preserve_host = false, host change in plugin: upstream host", function()
```

Comment:
So, now to answer this question:

> Do you know of any gotchas that would make this behavior unwanted?

Which I was about to do in the previous patch, but then noted that it was because the tests were using `add` instead of `replace`.

To answer the question, I believe that it isn't an unreasonable assumption for a user to expect a plugin to override a Route's settings. As of this patch, we only allow replacing the Host header if `preserve_host = false` (which, arguable is even more confusing that the previous assumed behavior - which was incorrectly described, but that is of no importance).

Additionally, there is a deeper issue at hand here I think, which is that the reason why this test case passes (and why the plugin cannot replace a Host header when `preserve_host = false`, is because we rely on `ngx.var.upstream_host`, which itself relies on the router's behavior. So the reason why this does not work (or why this test case succeeds) is because we directly rely on router behavior [see upstream_host is nil](https://github.com/Kong/kong/blob/master/kong/router.lua#L1159) which causes [ngx.var.upstream_host to be nil](https://github.com/Kong/kong/blob/master/kong/runloop/handler.lua#L751), which causes this plugin to not act on a `replace = { "host:foo" }` configuration value, because we guard the replace branch with `if headers[name] ~= nil then`.

In my opinion, this plugin could safely set the upstream header regardless of whether `preserve_host` is `true` or `false`. In doing so, we would avoid relying on core behavior as we currently do, which isn't regression-safe at all.
We should find another way than the `headers.host = ngx.var.upstream_host`. In my opinion, we are lacking a thin layer of abstraction between the core and this plugin (we may use existing PDK functions that act as this layer, or we may come up with a new PDK abstraction). This is one of these problems that #3821 is trying to solve, iirc.

Reply:
Yeah, I had a bad feeling about this particular case you highlighted as well. Since the original issue was about corrupting `Host` when transformations were not being applied on it and the semantics of the behavior of transforming `Host` were unclear anyway, I pushed on with that original approach.

> In my opinion, this plugin could safely set the upstream header regardless of whether preserve_host is true or false.

I like these simpler semantics. 

> In doing so, we would avoid relying on core behavior as we currently do, which isn't regression-safe at all.

In the latest iteration, I had to special-case the `Host` header a bit more in the plugin again, but it does not rely on core behavior (only on the assumptions that a `Host` header must always exist [to forbid `add` and `rename`] and cannot be set multiple times [to forbid `append`] ). 

> In my opinion, we are lacking a thin layer of abstraction between the core and this plugin (we may use existing PDK functions that act as this layer, or we may come up with a new PDK abstraction). This is one of these problems that #3821 is trying to solve, iirc.

Agreed, because since `kong.request` represents the incoming request, its value of `get_header("host")` is correct. Perhaps what we really want is `kong.service.request.get_headers`, which would represent the state of headers as they are being prepared to be sent to the outgoing request (but of course there are difficulties in crafting such an API and keeping `ngx` and the PDK interoperable).

I think for now the approach in this PR is sufficient to fix the related issue.

Reply:
> In the latest iteration, I had to special-case the Host header a bit more in the plugin again, but it does not rely on core behavior (only on the assumptions that a Host header must always exist [to forbid add and rename] and cannot be set multiple times [to forbid append] ).

This iteration looks better in its behavior. But it's still somewhat hand-crafted based on deep Core behavior, and a thin abstraction would be nicer. Nonetheless, the patch looks good for now :+1:

> Perhaps what we really want is kong.service.request.get_headers, which would represent the state of headers as they are being prepared to be sent to the outgoing request

This has been proposed in #3840 already - there has been a few asks around it and it seems like a valuable addition to the PDK.

Reply:
Actually, I take that back; I think the current level of decoupling between the core and this plugin is fine as per this iteration :+1: